### DISCUSSION TITLE
How should we distribute piped agent?

### DISCUSSION THREAD

**Discussion by kentakozuka** [pipe-cd/pipecd#4380] [Ideas]
**Background**

Before the introduction of `CUSTOM_SYNC` stage (issue https://github.com/pipe-cd/pipecd/issues/1089), PipeCD only supported built-in platform providers.

We want to give users more freedom to sync/apply/reconcile their applications via `CUSTOM_SYNC` stage, in which they can use $SHELL commands to deploy scripts. With that change, we thought piped needed a way to manage which binary/commands are available and which are not, which led to the introduction of [`asdf`](https://asdf-vm.com/) as a binary management tool for all (ref: [docs dev](https://pipecd.dev/docs-dev/user-guide/managing-application/customizing-deployment/custom-sync/#external-tool-management)).

The problem is that in some cases, asdf only installs a binary and does not ensure the installed binary works (issue https://github.com/pipe-cd/pipecd/issues/4376).

Based on the fact that we should focus on the functionality of the piped agent while keeping the cost of maintaining manifests and dependencies as low as possible (simpler is better), we think instead of trying to find the best binary management which fits all the cases, we should keep the focus on the piped running functional and lets users handle the environment.

**Ideas**

- Core team provides binaries and minimal images only
- The community will create the images to be able to use sam commands, etc.
- Documentation should clearly state how to use custom images.
- Remove asdf
- Clean up piped-base. Remove INSTALL for each command

**Comment by kentakozuka** [comment]
- Community members contribute to a repository for creating custom images for Custom Sync.

**Comment by kentakozuka** [comment]
Just thought;
We simply create asdf plugin for the requirements of Python installation.
@khanhtc1202 wdyt?

Though I am not sure it would be possible...

**Reply by khanhtc1202** [reply]
That's a point. If we choose to go that way, we have to keep supporting asdf plugins like that. And besides, even if we have a way to install Python using asdf, as the example, the sam-cli users have to know and configure the piped/application externalTools to install python then sam-cli which is not very natural to me.

I mean, compare installing all required commands via declarable Dockerfile vs pipecd asdf plugin management python then sam-cli, I think the first way is easier to understand.

**Reply by kentakozuka** [reply]
That's true 👍 
Let's go a way to the first one 🚀

**Comment by kentakozuka** [comment]
What is the best name for a repository where the community can maintain piped agent images?

**Reply by kentakozuka** [reply]
What about `piped-images`?

**Reply by khanhtc1202** [reply]
@kentakozuka I think that name is a bit confusing, but I don't have a better one now as well 😓 Lets use that name and add description to its readme 👍

**Comment by khanhtc1202** [comment]
Update:
- We removed the asdf installation/usage from the pipecd project (via https://github.com/pipe-cd/pipecd/pull/4383)
- Remove all external tools installed by default from the piped-based image (via https://github.com/pipe-cd/pipecd/pull/4400)

**Comment by ffjlabo** [comment]
We will discuss about preparing the tools by container image later 👍
